File: src/main/scala/jk_5/nailed/server/mappack/metadata/xml/XmlMappackMetadata.scala
/*
 * Nailed, a Minecraft PvP server framework
 * Copyright (C) jk-5 <http://github.com/jk-5/>
 * Copyright (C) Nailed team and contributors <http://github.com/nailed/>
 *
 * This program is free software: you can redistribute it and/or modify it
 * under the terms of the MIT License.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License
 * for more details.
 *
 * You should have received a copy of the MIT License along with
 * this program. If not, see <http://opensource.org/licenses/MIT/>.
 */

package jk_5.nailed.server.mappack.metadata.xml

import java.io.File

import jk_5.nailed.api.mappack._
import jk_5.nailed.api.mappack.metadata._
import jk_5.nailed.api.mappack.metadata.impl.DefaultMappackAuthor
import org.jdom2.input.SAXBuilder
import org.jdom2.{Element, Namespace}

import scala.collection.convert.wrapAsScala._

/**
 * No description given
 *
 * @author jk-5
 */
object XmlMappackMetadata {

  final val builder = new SAXBuilder()

  def fromFile(file: File): XmlMappackMetadata = {
    val doc = builder.build(file)
    val e = doc.getRootElement
    if(e.getName == "game"){
      new XmlMappackMetadata(e, e.getNamespace)
    }else{
      throw new MappackConfigurationException("game.xml does not contain the root game element")
    }
  }
}

class XmlMappackMetadata(element: Element, ns: Namespace) extends MappackMetadata {
  if(element.getChild("name", ns) == null) throw new MappackConfigurationException("Missing required <name> element")
  if(element.getChild("version", ns) == null) throw new MappackConfigurationException("Missing required <version> element")
  if(element.getChild("authors", ns) == null) throw new MappackConfigurationException("Missing required <authors> element")

  override val name = element.getChild("name", ns).getText
  override val version = element.getChild("version", ns).getText

  override val authors: Array[MappackAuthor] = element.getChild("authors", ns).getChildren.map{ e =>
    if(e.getName != "author") throw new MappackConfigurationException("Invalid element in authors list: " + e.getName)
    if(e.getChild("name", ns) == null) throw new MappackConfigurationException("Missing required element <name> in <author> element")
    if(e.getChild("role", ns) == null) throw new MappackConfigurationException("Missing required element <role> in <author> element")
    new DefaultMappackAuthor(e.getChild("name", ns).getText, e.getChild("role", ns).getText)
  }.toArray

  override val tutorial = if(element.getChild("tutorial", ns) != null) new XmlTutorial(element.getChild("tutorial", ns)) else null
  override val teams: Array[MappackTeam] = if(element.getChild("teams", ns) == null) new Array[MappackTeam](0) else element.getChild("teams", ns).getChildren.map(e => new XmlMappackTeam(e)).toArray
  override val worlds: Array[MappackWorld] = element.getChild("worlds", ns).getChildren.map { e =>
    if(e.getChild("name", ns) == null) throw new MappackConfigurationException("Missing required element <name> in <world> element")
    new XmlMappackWorld(e.getChild("name", ns).getText, e)
  }.toArray
  override val gameType = if(element.getChild("gametype", ns) != null) element.getChild("gametype", ns).getAttributeValue("name", ns) else null
  override val stats: Array[StatConfig] = if(element.getChild("stats", ns) != null) element.getChild("stats", ns).getChildren.map{ e =>
    if(e.getName != "stat") throw new MappackConfigurationException("Invalid element in stats list: " + e.getName)
    val name = e.getAttributes.find(_.getName == "name")
    if(name.isEmpty || name.get.getValue.length == 0) throw new MappackConfigurationException("Missing required attribute 'name' in stat element")
    new XmlStatConfig(name.get.getValue, e)
  }.toArray else new Array[StatConfig](0)
}

File: src/main/java/jk_5/nailed/server/mappack/metadata/xml/XmlMappackMetadata.java
package jk_5.nailed.server.mappack.metadata.xml;

import jk_5.nailed.api.mappack.MappackConfigurationException;
import jk_5.nailed.api.mappack.metadata.*;
import jk_5.nailed.api.mappack.metadata.impl.DefaultMappackAuthor;
import org.jdom2.Document;
import org.jdom2.Element;
import org.jdom2.JDOMException;
import org.jdom2.Namespace;
import org.jdom2.input.SAXBuilder;

import javax.annotation.Nonnull;
import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class XmlMappackMetadata implements MappackMetadata {

    private static final SAXBuilder builder = new SAXBuilder();

    private final String name;
    private final String version;
    private final MappackAuthor[] authors;
    private final XmlTutorial tutorial;
    private final MappackTeam[] teams;
    private final MappackWorld[] worlds;
    private final String gameType;
    private final StatConfig[] stats;

    public static XmlMappackMetadata fromFile(File file) throws MappackConfigurationException {
        try {
            Document doc = builder.build(file);
            Element e = doc.getRootElement();
            if(e.getName().equals("game")){
                return new XmlMappackMetadata(e, e.getNamespace());
            }else{
                throw new MappackConfigurationException("game.xml does not contain the root game element");
            }
        }catch(JDOMException e){
            throw new MappackConfigurationException("game.xml syntax error", e);
        }catch(IOException e){
            throw new MappackConfigurationException("Was not able to read game.xml", e);
        }
    }

    public static XmlMappackMetadata fromResource(String path) throws MappackConfigurationException {
        try {
            Document doc = builder.build(XmlMappackMetadata.class.getResourceAsStream("/" + path));
            Element e = doc.getRootElement();
            if(e.getName().equals("game")){
                return new XmlMappackMetadata(e, e.getNamespace());
            }else{
                throw new MappackConfigurationException("game.xml does not contain the root game element");
            }
        }catch(JDOMException e){
            throw new MappackConfigurationException("game.xml syntax error", e);
        }catch(IOException e){
            throw new MappackConfigurationException("Was not able to read game.xml", e);
        }
    }

    public XmlMappackMetadata(Element element, Namespace ns) throws MappackConfigurationException {
        if(element.getChild("name", ns) == null){
            throw new MappackConfigurationException("Missing required <name> element");
        }
        if(element.getChild("version", ns) == null){
            throw new MappackConfigurationException("Missing required <version> element");
        }
        if(element.getChild("authors", ns) == null){
            throw new MappackConfigurationException("Missing required <authors> element");
        }
        if(element.getChild("worlds", ns) == null){
            throw new MappackConfigurationException("Missing required <worlds> element");
        }
        if(element.getChild("worlds", ns).getChildren().size() == 0){
            throw new MappackConfigurationException("<worlds> element should have at least 1 <world>");
        }

        this.name = element.getChild("name", ns).getText();
        this.version = element.getChild("version", ns).getText();

        List<MappackAuthor> authors = new ArrayList<MappackAuthor>();
        for (Element e : element.getChild("authors", ns).getChildren()) {
            if(!e.getName().equals("author")){
                throw new MappackConfigurationException("Invalid element in authors list: " + e.getName());
            }
            if(e.getChild("name", ns) == null){
                throw new MappackConfigurationException("Missing required element <name> in <author> element");
            }
            if(e.getChild("role", ns) == null){
                throw new MappackConfigurationException("Missing required element <role> in <author> element");
            }
            authors.add(new DefaultMappackAuthor(e.getChild("name", ns).getText(), e.getChild("role", ns).getText()));
        }

        this.authors = authors.toArray(new MappackAuthor[authors.size()]);

        if(element.getChild("tutorial", ns) == null){
            this.tutorial = null;
        }else{
            this.tutorial = new XmlTutorial(element.getChild("tutorial", ns));
        }

        if(element.getChild("teams", ns) == null){
            this.teams = new MappackTeam[0];
        }else{
            List<MappackTeam> teams = new ArrayList<MappackTeam>();
            for (Element e : element.getChild("teams", ns).getChildren()) {
                teams.add(new XmlMappackTeam(e));
            }
            this.teams = teams.toArray(new MappackTeam[teams.size()]);
        }

        List<MappackWorld> worlds = new ArrayList<MappackWorld>();
        for (Element e : element.getChild("worlds", ns).getChildren()) {
            if(e.getChild("name", ns) == null){
                throw new MappackConfigurationException("Missing required element <name> in <world> element");
            }
            worlds.add(new XmlMappackWorld(e.getChild("name", ns).getText(), e, null)); //TODO: remove null in here as it is overloaded
        }
        this.worlds = worlds.toArray(new MappackWorld[worlds.size()]);

        if(element.getChild("gametype", ns) == null){
            this.gameType = null;
        }else{
            this.gameType = element.getChild("gametype").getAttributeValue("name", ns);
        }

        if(element.getChild("stats", ns) == null){
            this.stats = new StatConfig[0];
        }else{
            List<StatConfig> stats = new ArrayList<StatConfig>();
            for (Element e : element.getChild("stats", ns).getChildren()) {
                if(!e.getName().equals("stat")){
                    throw new MappackConfigurationException("Invalid element in stats list: " + e.getName());
                }
                if(e.getAttribute("name", ns) == null){
                    throw new MappackConfigurationException("Missing required attribute 'name' in stat element");
                }
                String name = e.getAttribute("name", ns).getValue();
                if(name.isEmpty() || name.length() == 0){
                    throw new MappackConfigurationException("Missing required attribute 'name' in stat element");
                }
                stats.add(new XmlStatConfig(name, e));
            }
            this.stats = stats.toArray(new StatConfig[stats.size()]);
        }
    }

    @Nonnull
    @Override
    public String name() {
        return this.name;
    }

    @Nonnull
    @Override
    public String version() {
        return this.version;
    }

    @Nonnull
    @Override
    public MappackAuthor[] authors() {
        return this.authors;
    }

    @Nonnull
    @Override
    public MappackWorld[] worlds() {
        return this.worlds;
    }

    @Nonnull
    @Override
    public MappackTeam[] teams() {
        return this.teams;
    }

    @Nonnull
    @Override
    public Tutorial tutorial() {
        return this.tutorial;
    }

    @Nonnull
    @Override
    public String gameType() {
        return this.gameType;
    }

    @Nonnull
    @Override
    public StatConfig[] stats() {
        return this.stats;
    }
}
